File: swift_mort/MORT/Models/MissionPilot.swift
import Foundation

struct ClosedPilotEligibility: Codable, Hashable, Sendable {
    let allowed: Bool
    let code: String?
    let missingRequirements: [String]?
    let reasonCodes: [String]?
    let policyVersion: Int?
    let pilotModeEnabled: Bool?
    let unrestrictedPublicAccessEnabled: Bool?
    let realDocumentCollectionEnabled: Bool?
    let guardianModeOptional: Bool?
    let guardianConnectionRequired: Bool?
    let permanentAddressRequired: Bool?
    let housingStatusCollected: Bool?
    let supportCircleAffectsEligibility: Bool?

    enum CodingKeys: String, CodingKey {
        case allowed, code
        case missingRequirements = "missing_requirements"
        case reasonCodes = "reason_codes"
        case policyVersion = "policy_version"
        case pilotModeEnabled = "pilot_mode_enabled"
        case unrestrictedPublicAccessEnabled = "unrestricted_public_access_enabled"
        case realDocumentCollectionEnabled = "real_document_collection_enabled"
        case guardianModeOptional = "guardian_mode_optional"
        case guardianConnectionRequired = "guardian_connection_required"
        case permanentAddressRequired = "permanent_address_required"
        case housingStatusCollected = "housing_status_collected"
        case supportCircleAffectsEligibility = "support_circle_affects_eligibility"
    }
}

struct MissionPilotDashboard: Codable, Hashable, Sendable {
    let ok: Bool
    let mission: String
    let pilotEligibility: ClosedPilotEligibility
    let discreetMode: DiscreetModeState
    let supportCircle: SupportCircleState
    let activeGoalCount: Int
    let reviewedResourceCount: Int
    let documentReview: DocumentCollectionReadiness
    let mortHoldsPayments: Bool
    let runawayGuidanceProvided: Bool

    enum CodingKeys: String, CodingKey {
        case ok, mission
        case pilotEligibility = "pilot_eligibility"
        case discreetMode = "discreet_mode"
        case supportCircle = "support_circle"
        case activeGoalCount = "active_goal_count"
        case reviewedResourceCount = "reviewed_resource_count"
        case documentReview = "document_review"
        case mortHoldsPayments = "mort_holds_payments"
        case runawayGuidanceProvided = "runaway_guidance_provided"
    }
}

struct DiscreetModeState: Codable, Hashable, Sendable {
    let enabled: Bool
    let genericNotificationTitle: Bool?
    let hideNotificationContent: Bool?
    let appLockEnabled: Bool?
    let automaticLockMinutes: Int?
    let quickExitDestination: String?

    enum CodingKeys: String, CodingKey {
        case enabled
        case genericNotificationTitle = "generic_notification_title"
        case hideNotificationContent = "hide_notification_content"
        case appLockEnabled = "app_lock_enabled"
        case automaticLockMinutes = "automatic_lock_minutes"
        case quickExitDestination = "quick_exit_destination"
    }
}

struct SupportCircleState: Codable, Hashable, Sendable {
    let enabled: Bool
    let memberCount: Int
    let optional: Bool
    let affectsEligibility: Bool

    enum CodingKeys: String, CodingKey {
        case enabled, optional
        case memberCount = "member_count"
        case affectsEligibility = "affects_eligibility"
    }
}

struct DocumentCollectionReadiness: Codable, Hashable, Sendable {
    let ready: Bool
    let realDocumentCollectionEnabled: Bool
    let clientCanEnable: Bool
    let collectionStatus: String
    let requiredGateCount: Int
    let passedGateCount: Int
    let remainingGateKeys: [String]
    let truthStatement: String

    enum CodingKeys: String, CodingKey {
        case ready
        case realDocumentCollectionEnabled = "real_document_collection_enabled"
        case clientCanEnable = "client_can_enable"
        case collectionStatus = "collection_status"
        case requiredGateCount = "required_gate_count"
        case passedGateCount = "passed_gate_count"
        case remainingGateKeys = "remaining_gate_keys"
        case truthStatement = "truth_statement"
    }
}

struct MissionPilotActionResult: Codable, Hashable, Sendable {
    let ok: Bool
    let code: String?
    let status: String?
    let message: String?
    let enrollmentID: UUID?
    let manualApprovalRequired: Bool?
    let guardianModeOptional: Bool?
    let permanentAddressRequired: Bool?
    let enabled: Bool?
    let optional: Bool?
    let affectsMarketplaceEligibility: Bool?

    enum CodingKeys: String, CodingKey {
        case ok, code, status, message, enabled, optional
        case enrollmentID = "enrollment_id"
        case manualApprovalRequired = "manual_approval_required"
        case guardianModeOptional = "guardian_mode_optional"
        case permanentAddressRequired = "permanent_address_required"
        case affectsMarketplaceEligibility = "affects_marketplace_eligibility"
    }
}

struct PartnerAttestationItem: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    let factType: String
    let version: Int
    let status: String
    let statement: String
    let whatWasNotEstablished: String
    let effectiveAt: String
    let expiresAt: String

    enum CodingKeys: String, CodingKey {
        case id, version, status, statement
        case factType = "fact_type"
        case whatWasNotEstablished = "what_was_not_established"
        case effectiveAt = "effective_at"
        case expiresAt = "expires_at"
    }
}

struct PartnerAttestationsResponse: Codable, Hashable, Sendable {
    let ok: Bool
    let identityVerified: Bool
    let governmentIDVerified: Bool
    let attestations: [PartnerAttestationItem]

    enum CodingKeys: String, CodingKey {
        case ok, attestations
        case identityVerified = "identity_verified"
        case governmentIDVerified = "government_id_verified"
    }
}

struct DocumentReviewCaseSummary: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    let evidenceCategory: String
    let status: String
    let publicLabel: String
    let whatWasEstablished: String
    let whatWasNotEstablished: String
    let finalDecisionAt: String?
    let expiresAt: String?

    enum CodingKeys: String, CodingKey {
        case id, status
        case evidenceCategory = "evidence_category"
        case publicLabel = "public_label"
        case whatWasEstablished = "what_was_established"
        case whatWasNotEstablished = "what_was_not_established"
        case finalDecisionAt = "final_decision_at"
        case expiresAt = "expires_at"
    }
}

struct IndependenceGoal: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    let goalType: String
    let title: String
    let targetAmountCents: Int?
    let currentAmountCents: Int
    let targetDate: String?
    let status: String

    enum CodingKeys: String, CodingKey {
        case id, title, status
        case goalType = "goal_type"
        case targetAmountCents = "target_amount_cents"
        case currentAmountCents = "current_amount_cents"
        case targetDate = "target_date"
    }
}

struct ResourceDirectoryEntry: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    let organizationName: String
    let category: String
    let sourceURL: String
    let sourceStatus: String
    let organizationVerificationStatus: String
    let city: String?
    let state: String?
    let summary: String
    let emergencyLimitations: String
    let availabilityClaimed: Bool

    enum CodingKeys: String, CodingKey {
        case id, category, city, state, summary
        case organizationName = "organization_name"
        case sourceURL = "source_url"
        case sourceStatus = "source_status"
        case organizationVerificationStatus = "organization_verification_status"
        case emergencyLimitations = "emergency_limitations"
        case availabilityClaimed = "availability_claimed"
    }
}

struct PrivateWorkSummary: Codable, Hashable, Sendable {
    let ok: Bool
    let totalSelfRecordedEarningsCents: Int
    let completedJobCount: Int
    let skillCount: Int
    let referenceCount: Int
    let mortHoldsPayments: Bool
    let mortGuaranteesPayments: Bool
    let privateByDefault: Bool

    enum CodingKeys: String, CodingKey {
        case ok
        case totalSelfRecordedEarningsCents = "total_self_recorded_earnings_cents"
        case completedJobCount = "completed_job_count"
        case skillCount = "skill_count"
        case referenceCount = "reference_count"
        case mortHoldsPayments = "mort_holds_payments"
        case mortGuaranteesPayments = "mort_guarantees_payments"
        case privateByDefault = "private_by_default"
    }
}

File: flutter_mort/lib/data/models/mission_pilot.dart
class ClosedPilotEligibility {
  const ClosedPilotEligibility({
    required this.allowed,
    required this.code,
    required this.missingRequirements,
    required this.reasonCodes,
    required this.guardianModeOptional,
    required this.permanentAddressRequired,
    required this.realDocumentCollectionEnabled,
  });

  final bool allowed;
  final String? code;
  final List<String> missingRequirements;
  final List<String> reasonCodes;
  final bool guardianModeOptional;
  final bool permanentAddressRequired;
  final bool realDocumentCollectionEnabled;

  factory ClosedPilotEligibility.fromMap(Map<String, dynamic> map) {
    return ClosedPilotEligibility(
      allowed: map['allowed'] == true,
      code: map['code']?.toString(),
      missingRequirements: _strings(map['missing_requirements']),
      reasonCodes: _strings(map['reason_codes']),
      guardianModeOptional: map['guardian_mode_optional'] != false,
      permanentAddressRequired: map['permanent_address_required'] == true,
      realDocumentCollectionEnabled:
          map['real_document_collection_enabled'] == true,
    );
  }
}

class MissionPilotDashboard {
  const MissionPilotDashboard({
    required this.mission,
    required this.eligibility,
    required this.discreetMode,
    required this.supportCircle,
    required this.activeGoalCount,
    required this.reviewedResourceCount,
    required this.documentReadiness,
  });

  final String mission;
  final ClosedPilotEligibility eligibility;
  final Map<String, dynamic> discreetMode;
  final Map<String, dynamic> supportCircle;
  final int activeGoalCount;
  final int reviewedResourceCount;
  final DocumentCollectionReadiness documentReadiness;

  factory MissionPilotDashboard.fromMap(Map<String, dynamic> map) {
    return MissionPilotDashboard(
      mission:
          map['mission']?.toString() ??
          'Help teenagers gain legitimate income, work experience, skills, references, and safe pathways toward adulthood.',
      eligibility: ClosedPilotEligibility.fromMap(
        _map(map['pilot_eligibility']),
      ),
      discreetMode: _map(map['discreet_mode']),
      supportCircle: _map(map['support_circle']),
      activeGoalCount: _integer(map['active_goal_count']),
      reviewedResourceCount: _integer(map['reviewed_resource_count']),
      documentReadiness: DocumentCollectionReadiness.fromMap(
        _map(map['document_review']),
      ),
    );
  }
}

class PartnerAttestation {
  const PartnerAttestation({
    required this.id,
    required this.factType,
    required this.version,
    required this.status,
    required this.statement,
    required this.whatWasNotEstablished,
  });

  final String id;
  final String factType;
  final int version;
  final String status;
  final String statement;
  final String whatWasNotEstablished;

  factory PartnerAttestation.fromMap(Map<String, dynamic> map) {
    return PartnerAttestation(
      id: map['id'].toString(),
      factType: map['fact_type']?.toString() ?? 'partner_relationship',
      version: _integer(map['version']),
      status: map['status']?.toString() ?? 'unknown',
      statement: map['statement']?.toString() ?? 'No statement available.',
      whatWasNotEstablished:
          map['what_was_not_established']?.toString() ??
          'Government identity was not established.',
    );
  }
}

class DocumentCollectionReadiness {
  const DocumentCollectionReadiness({
    required this.ready,
    required this.realDocumentCollectionEnabled,
    required this.clientCanEnable,
    required this.requiredGateCount,
    required this.passedGateCount,
    required this.remainingGateKeys,
    required this.truthStatement,
  });

  final bool ready;
  final bool realDocumentCollectionEnabled;
  final bool clientCanEnable;
  final int requiredGateCount;
  final int passedGateCount;
  final List<String> remainingGateKeys;
  final String truthStatement;

  factory DocumentCollectionReadiness.fromMap(Map<String, dynamic> map) {
    return DocumentCollectionReadiness(
      ready: map['ready'] == true,
      realDocumentCollectionEnabled:
          map['real_document_collection_enabled'] == true,
      clientCanEnable: map['client_can_enable'] == true,
      requiredGateCount: _integer(map['required_gate_count']),
      passedGateCount: _integer(map['passed_gate_count']),
      remainingGateKeys: _strings(map['remaining_gate_keys']),
      truthStatement:
          map['truth_statement']?.toString() ??
          'Visual review does not by itself prove document authenticity or legal identity.',
    );
  }
}

class IndependenceGoal {
  const IndependenceGoal({
    required this.id,
    required this.goalType,
    required this.title,
    required this.targetAmountCents,
    required this.currentAmountCents,
    required this.status,
  });

  final String id;
  final String goalType;
  final String title;
  final int? targetAmountCents;
  final int currentAmountCents;
  final String status;

  factory IndependenceGoal.fromMap(Map<String, dynamic> map) {
    return IndependenceGoal(
      id: map['id'].toString(),
      goalType: map['goal_type']?.toString() ?? 'custom',
      title: map['title']?.toString() ?? 'Private goal',
      targetAmountCents: map['target_amount_cents'] == null
          ? null
          : _integer(map['target_amount_cents']),
      currentAmountCents: _integer(map['current_amount_cents']),
      status: map['status']?.toString() ?? 'active',
    );
  }
}

class ResourceDirectoryEntry {
  const ResourceDirectoryEntry({
    required this.id,
    required this.organizationName,
    required this.category,
    required this.sourceUrl,
    required this.sourceStatus,
    required this.summary,
    required this.emergencyLimitations,
  });

  final String id;
  final String organizationName;
  final String category;
  final String sourceUrl;
  final String sourceStatus;
  final String summary;
  final String emergencyLimitations;

  factory ResourceDirectoryEntry.fromMap(Map<String, dynamic> map) {
    return ResourceDirectoryEntry(
      id: map['id'].toString(),
      organizationName: map['organization_name']?.toString() ?? 'Resource',
      category: map['category']?.toString() ?? 'community_organization',
      sourceUrl: map['source_url']?.toString() ?? '',
      sourceStatus: map['source_status']?.toString() ?? 'pending_review',
      summary: map['summary']?.toString() ?? '',
      emergencyLimitations:
          map['emergency_limitations']?.toString() ??
          'This listing does not replace emergency services.',
    );
  }
}

Map<String, dynamic> missionMap(Object? value) => _map(value);

Map<String, dynamic> _map(Object? value) {
  if (value is Map<String, dynamic>) return value;
  if (value is Map) return Map<String, dynamic>.from(value);
  return <String, dynamic>{};
}

List<String> _strings(Object? value) {
  if (value is! List) return const [];
  return value.map((item) => item.toString()).toList(growable: false);
}

int _integer(Object? value) {
  if (value is int) return value;
  return int.tryParse(value?.toString() ?? '') ?? 0;
}
